docs/graphql/manual/queries/performance.rst

```diff
@@ -0,0 +1,238 @@
+
+1. The prepared statement
+2. Information necessary to convert variables into the prepared statement's arguments
+
+For the above query, Hasura generates the following prepared statement (simplified):
+
+.. code-block:: plpgsql
+
+   select name, rating from author where id = $1
+
+With the following prepared variables:
+
+.. code-block:: plpgsql
+
+   $1 = 1
+
+This plan is then saved in a data structure called ``Query Plan Cache``. The next time the same query is executed, 
+Hasura uses the plan to convert the provided variables into the prepared statement's arguments and then executes the statement. 
+This will significantly cut down the execution time for a GraphQL query resulting in lower latencies and higher throughput.
+
+Caveats
+^^^^^^^
+
+The above optimization is not possible for all types of queries. For example, consider this query:
+
+.. code-block:: graphql
+
+   query getAuthorWithCondition($condition: author_bool_exp!) {
+      author(where: $condition)
+         name
+         rating
+      }
+   }
+
+The statement generated for ``getAuthorWithCondition`` is now dependent on the variables.
+
+With the following variables:
+
+.. code-block:: json
+
+   {
+      "condition": {"id": {"_eq": 1}}
+   }
+
+the generated statement will be:
+
+.. code-block:: plpgsql
+
+   select name, rating from author where id = $1
+
+However, with the following variables:
+
+.. code-block:: json
+
+   {
+      "condition": {"name": {"_eq": "John"}}
+   }
+
+the generated statement will be:
+
+.. code-block:: plpgsql
+
+   select name, rating from author where name = 'John'
+
+A plan cannot be generated for such queries because the variables defined in the GraphQL query don't have a one-to-one correspondence to the parameters in the prepared statement.
+
+Query optimization
+------------------
+
+Using GraphQL variables
+^^^^^^^^^^^^^^^^^^^^^^^
+
+In order to leverage Hasura's query plan caching (as explained in the :ref:`previous section <query_plan_caching>`) to the full extent, GraphQL queries should be defined with
+variables whose types are **non-nullable** whenever possible.
+
+To make variables non-nullable, add a ``!`` at the end of the type, like here:
+
+.. code-block:: graphql
+   :emphasize-lines: 1
+
+   query getAuthor($id: Int!) {
+      authors(where: {id: {_eq: $id}}) {
+         name
+         rating
+      }
+   }
+
+If the ``!`` is not added and the variable is nullable, the generated query will be different depending if an ``id`` is passed or if the variables is ``null``
+(for the latter, there is no ``where`` statement present). Therefore, it's not possible for Hasura to create a reusable plan for a query in this case.
+
+.. note::
+
+   Hasura is fast even for queries which cannot have a reusable plan.
+   This should concern you only if you face a high volume of traffic (thousands of requests per second).
+
+Using PG indexes
```

Comment:
Why not move this section to immediately after Analyze?

Reply:
I think the intent was to make sure someone reading the query optimisation section specifically is made aware of this fact. The earlier section is more of a reference that ppl might skip